plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java

```diff
@@ -231,6 +232,42 @@ public static Table getIcebergTableWithMetadata(
         return new BaseTable(operations, quotedTableName(table), TRINO_METRICS_REPORTER);
     }
 
+    public static List<IcebergColumnHandle> getProjectedColumns(Schema schema, TypeManager typeManager)
+    {
+        ImmutableList.Builder<IcebergColumnHandle> projectedColumns = ImmutableList.builder();
+        StructType schemaAsStruct = schema.asStruct();
+        Map<Integer, NestedField> indexById = TypeUtil.indexById(schemaAsStruct);
+        Map<Integer, Integer> indexParents = TypeUtil.indexParents(schemaAsStruct);
+        Map<Integer, List<Integer>> indexPaths = indexById.entrySet().stream()
+                .collect(toImmutableMap(Entry::getKey, e -> buildPath(indexParents, e.getKey())));
+
+        for (Map.Entry<Integer, NestedField> entry : indexById.entrySet()) {
+            int fieldId = entry.getKey();
+            NestedField childField = entry.getValue();
+            NestedField baseField = childField;
+
+            List<Integer> path = new ArrayList<>(requireNonNull(indexPaths.get(fieldId)));
+            if (!path.isEmpty()) {
+                // Path does not include the base field id
+                baseField = indexById.get(path.removeFirst());
```

Comment:
It is hard to follow what is happening when we have mutations.

Why do we remove the first element from the list?

Reply:
The first element is used for getting the root field id, but the element shouldn't exist for `IcebergColumnHandle.path`. I updated the comment and slightly modified the logic.